File: kb-core-ui/internal/bots/runner.go
package bots

import (
	"bytes"
	"os/exec"
	"sort"
	"strconv"
	"strings"
	"sync"
	"time"
)

// Run is a single execution of a bot (mirrors API_CONTRACT.md's BotRun).
type Run struct {
	ID         string     `json:"id"`
	Bot        string     `json:"bot"`
	Status     string     `json:"status"` // "running" | "succeeded" | "failed"
	StartedAt  time.Time  `json:"startedAt"`
	FinishedAt *time.Time `json:"finishedAt,omitempty"`
	ExitCode   *int       `json:"exitCode,omitempty"`
	Output     string     `json:"output"`
}

// RunSummary is a Run without its (potentially large) output, for list views.
type RunSummary struct {
	ID         string     `json:"id"`
	Bot        string     `json:"bot"`
	Status     string     `json:"status"`
	StartedAt  time.Time  `json:"startedAt"`
	FinishedAt *time.Time `json:"finishedAt,omitempty"`
	ExitCode   *int       `json:"exitCode,omitempty"`
}

// Runner executes bots by re-invoking the kb-core-ui binary and tracks their
// runs in memory. Safe for concurrent use. Runs are not persisted — this is
// a local dev tool; restarting kb-core-ui clears run history.
type Runner struct {
	selfBin  string // path to the kb-core-ui binary to re-invoke
	repoRoot string

	mu   sync.Mutex
	runs map[string]*Run
	seq  int
	now  func() time.Time // injectable for tests
}

// NewRunner builds a Runner. selfBin is the kb-core-ui executable to invoke
// (`<selfBin> bot <sub> ...`); repoRoot is the repo bots operate on.
func NewRunner(selfBin, repoRoot string) *Runner {
	return &Runner{
		selfBin:  selfBin,
		repoRoot: repoRoot,
		runs:     make(map[string]*Run),
		now:      time.Now,
	}
}

// ErrUnknownBot / ErrMissingArg are returned by Start for bad requests so
// the HTTP layer can map them to 404/400.
type ErrUnknownBot struct{ Name string }

func (e ErrUnknownBot) Error() string { return "unknown bot: " + e.Name }

type ErrMissingArg struct{ Arg string }

func (e ErrMissingArg) Error() string { return "missing required arg: " + e.Arg }

// Start validates args, launches the bot in the background, and returns the
// initial (running) Run immediately. Output accumulates on the Run until it
// finishes; callers poll Get to observe progress.
func (r *Runner) Start(botName string, args map[string]string) (*Run, error) {
	def, ok := Lookup(botName)
	if !ok {
		return nil, ErrUnknownBot{botName}
	}
	for _, a := range def.Args {
		if a.Required && args[a.Name] == "" {
			return nil, ErrMissingArg{a.Name}
		}
	}

	cmdArgs, err := buildArgs(def, args, r.repoRoot)
	if err != nil {
		return nil, err
	}

	r.mu.Lock()
	r.seq++
	id := "run-" + strconv.Itoa(r.seq)
	run := &Run{
		ID:        id,
		Bot:       botName,
		Status:    "running",
		StartedAt: r.now(),
	}
	r.runs[id] = run
	r.mu.Unlock()

	go r.execute(run, cmdArgs)
	return r.snapshot(id), nil
}

// buildArgs turns a bot def + user args into the argv after
// `kb-core-ui bot`. Positional args (declared in def.positional) come first
// in order; the rest become flags — boolean args as "--name" when truthy,
// others as "--name value". Arg names use underscores (JSON-friendly) but
// CLI flags use dashes, so "dry_run" -> "--dry-run". Every bot also gets
// --repo <repoRoot> so it operates on the served repo.
func buildArgs(def Def, args map[string]string, repoRoot string) ([]string, error) {
	byName := map[string]ArgDef{}
	for _, a := range def.Args {
		byName[a.Name] = a
	}

	out := []string{"bot", def.subcommand}
	posSet := map[string]bool{}
	for _, name := range def.positional {
		posSet[name] = true
		out = append(out, args[name]) // positional value (may be empty for optional ones)
	}

	// Deterministic flag order for testability.
	var flagNames []string
	for _, a := range def.Args {
		if posSet[a.Name] || args[a.Name] == "" {
			continue
		}
		flagNames = append(flagNames, a.Name)
	}
	sort.Strings(flagNames)
	for _, name := range flagNames {
		flag := "--" + strings.ReplaceAll(name, "_", "-")
		if byName[name].Bool {
			if isTruthy(args[name]) {
				out = append(out, flag)
			}
			continue
		}
		out = append(out, flag, args[name])
	}
	out = append(out, "--repo", repoRoot)
	return out, nil
}

func isTruthy(s string) bool {
	switch s {
	case "true", "1", "yes", "on":
		return true
	}
	return false
}

func (r *Runner) execute(run *Run, cmdArgs []string) {
	cmd := exec.Command(r.selfBin, cmdArgs...)
	cmd.Dir = r.repoRoot

	// One buffer for combined stdout+stderr, guarded so the polling reader
	// (Get) sees a consistent snapshot while the process writes.
	lw := &lockedWriter{}
	cmd.Stdout = lw
	cmd.Stderr = lw

	// Stream output into the Run as it's produced.
	stop := make(chan struct{})
	go func() {
		ticker := time.NewTicker(200 * time.Millisecond)
		defer ticker.Stop()
		for {
			select {
			case <-ticker.C:
				r.mu.Lock()
				run.Output = lw.String()
				r.mu.Unlock()
			case <-stop:
				return
			}
		}
	}()

	err := cmd.Run()
	close(stop)

	exit := 0
	if err != nil {
		if ee, ok := err.(*exec.ExitError); ok {
			exit = ee.ExitCode()
		} else {
			exit = -1
			lw.WriteString("\n[runner] failed to execute bot: " + err.Error() + "\n")
		}
	}

	fin := r.now()
	r.mu.Lock()
	run.Output = lw.String()
	run.ExitCode = &exit
	run.FinishedAt = &fin
	if exit == 0 {
		run.Status = "succeeded"
	} else {
		run.Status = "failed"
	}
	r.mu.Unlock()
}

// Get returns a snapshot copy of one run, or ok=false.
func (r *Runner) Get(id string) (*Run, bool) {
	r.mu.Lock()
	defer r.mu.Unlock()
	run, ok := r.runs[id]
	if !ok {
		return nil, false
	}
	return copyRun(run), true
}

func (r *Runner) snapshot(id string) *Run {
	r.mu.Lock()
	defer r.mu.Unlock()
	return copyRun(r.runs[id])
}

// List returns run summaries newest-first.
func (r *Runner) List() []RunSummary {
	r.mu.Lock()
	defer r.mu.Unlock()
	out := make([]RunSummary, 0, len(r.runs))
	for _, run := range r.runs {
		out = append(out, RunSummary{
			ID: run.ID, Bot: run.Bot, Status: run.Status,
			StartedAt: run.StartedAt, FinishedAt: run.FinishedAt, ExitCode: run.ExitCode,
		})
	}
	sort.Slice(out, func(i, j int) bool { return out[i].StartedAt.After(out[j].StartedAt) })
	return out
}

func copyRun(run *Run) *Run {
	c := *run
	if run.FinishedAt != nil {
		t := *run.FinishedAt
		c.FinishedAt = &t
	}
	if run.ExitCode != nil {
		e := *run.ExitCode
		c.ExitCode = &e
	}
	return &c
}

// lockedWriter is an io.Writer + String() accumulator safe for the writer
// goroutine (cmd) and reader goroutine (poller) to share.
type lockedWriter struct {
	mu  sync.Mutex
	buf bytes.Buffer
}

func (w *lockedWriter) Write(p []byte) (int, error) {
	w.mu.Lock()
	defer w.mu.Unlock()
	return w.buf.Write(p)
}

func (w *lockedWriter) WriteString(s string) {
	w.mu.Lock()
	defer w.mu.Unlock()
	w.buf.WriteString(s)
}

func (w *lockedWriter) String() string {
	w.mu.Lock()
	defer w.mu.Unlock()
	return w.buf.String()
}

File: kb-core-ui/python/kb_core_ui/bots/runner.py
"""Executes bots by re-invoking the kb-core-ui entrypoint and tracks their
runs in memory — the Python side of internal/bots/runner.go.

Runs are not persisted: this is a local dev tool, and restarting clears the
history.
"""

from __future__ import annotations

import subprocess
import threading
from dataclasses import dataclass, field

from kb_core_ui.bots.registry import Def, lookup
from kb_core_ui.gotime import GoTime, now


@dataclass
class Run:
    id: str = ""
    bot: str = ""
    status: str = "running"  # "running" | "succeeded" | "failed"
    started_at: GoTime = field(default_factory=lambda: GoTime(0))
    finished_at: GoTime | None = None
    exit_code: int | None = None
    output: str = ""

    def to_json_dict(self) -> dict:
        out: dict = {
            "id": self.id,
            "bot": self.bot,
            "status": self.status,
            "startedAt": self.started_at.format(),
        }
        if self.finished_at is not None:
            out["finishedAt"] = self.finished_at.format()
        if self.exit_code is not None:
            out["exitCode"] = self.exit_code
        out["output"] = self.output
        return out

    def summary_json_dict(self) -> dict:
        out = self.to_json_dict()
        out.pop("output")
        return out


# Kept as a distinct name so the REST layer can keep mirroring Go's two
# structs, even though the summary is just a Run without its output.
RunSummary = Run


class UnknownBotError(Exception):
    def __init__(self, name: str):
        super().__init__("unknown bot: " + name)
        self.name = name


class MissingArgError(Exception):
    def __init__(self, arg: str):
        super().__init__("missing required arg: " + arg)
        self.arg = arg


def is_truthy(s: str) -> bool:
    return s in ("true", "1", "yes", "on")


def build_args(definition: Def, args: dict[str, str], repo_root: str) -> list[str]:
    """Turns a bot def + user args into the argv after `kb-core-ui bot`.
    Positional args come first in declaration order; the rest become flags,
    booleans as a bare "--name" when truthy. Arg names use underscores
    (JSON-friendly) while CLI flags use dashes, so "dry_run" -> "--dry-run".
    """
    by_name = {a.name: a for a in definition.args}

    out = ["bot", definition.subcommand]
    for name in definition.positional:
        # May be empty for optional positionals; Go passes the empty string.
        out.append(args.get(name, ""))

    # Deterministic flag order for testability.
    flag_names = sorted(
        a.name
        for a in definition.args
        if a.name not in definition.positional and args.get(a.name, "")
    )
    for name in flag_names:
        flag = "--" + name.replace("_", "-")
        if by_name[name].is_bool:
            if is_truthy(args[name]):
                out.append(flag)
            continue
        out += [flag, args[name]]
    out += ["--repo", repo_root]
    return out


class Runner:
    """Safe for concurrent use. self_argv is the command prefix that runs this
    same program (`<self_argv> bot <sub> ...`)."""

    def __init__(self, self_argv: list[str] | str, repo_root: str):
        self.self_argv = [self_argv] if isinstance(self_argv, str) else list(self_argv)
        self.repo_root = repo_root
        self._lock = threading.Lock()
        self._runs: dict[str, Run] = {}
        self._seq = 0
        self.now = now

    def start(self, bot_name: str, args: dict[str, str] | None) -> Run:
        """Validates args, launches the bot in the background, and returns the
        initial (running) Run. Callers poll get() to observe progress."""
        definition = lookup(bot_name)
        if definition is None:
            raise UnknownBotError(bot_name)
        args = args or {}
        for a in definition.args:
            if a.required and not args.get(a.name, ""):
                raise MissingArgError(a.name)

        cmd_args = build_args(definition, args, self.repo_root)

        with self._lock:
            self._seq += 1
            run = Run(id=f"run-{self._seq}", bot=bot_name, status="running",
                      started_at=self.now())
            self._runs[run.id] = run
            snapshot = _copy(run)

        threading.Thread(target=self._execute, args=(run, cmd_args), daemon=True).start()
        return snapshot

    def _execute(self, run: Run, cmd_args: list[str]) -> None:
        chunks: list[str] = []
        try:
            proc = subprocess.Popen(
                self.self_argv + cmd_args,
                cwd=self.repo_root,
                stdout=subprocess.PIPE,
                stderr=subprocess.STDOUT,
                text=True,
                encoding="utf-8",
                errors="replace",
            )
        except OSError as exc:
            with self._lock:
                run.output = f"\n[runner] failed to execute bot: {exc}\n"
                run.exit_code = -1
                run.finished_at = self.now()
                run.status = "failed"
            return

        assert proc.stdout is not None
        # Stream output into the Run as it is produced, so a polling reader
        # sees progress rather than only the final blob.
        for line in proc.stdout:
            chunks.append(line)
            with self._lock:
                run.output = "".join(chunks)
        exit_code = proc.wait()

        with self._lock:
            run.output = "".join(chunks)
            run.exit_code = exit_code
            run.finished_at = self.now()
            run.status = "succeeded" if exit_code == 0 else "failed"

    def get(self, run_id: str) -> Run | None:
        with self._lock:
            run = self._runs.get(run_id)
            return None if run is None else _copy(run)

    def list(self) -> list[Run]:
        with self._lock:
            runs = [_copy(r) for r in self._runs.values()]
        runs.sort(key=lambda r: r.started_at, reverse=True)
        return runs


def _copy(run: Run) -> Run:
    return Run(
        id=run.id,
        bot=run.bot,
        status=run.status,
        started_at=run.started_at,
        finished_at=run.finished_at,
        exit_code=run.exit_code,
        output=run.output,
    )
